sklearn/feature_extraction/text.py

```diff
@@ -1070,15 +1097,19 @@ def transform(self, X, copy=True):
             X.data += 1
 
         if self.use_idf:
-            check_is_fitted(self, '_idf_diag', 'idf vector is not fitted')
+            check_is_fitted(self, '_df', 'df vector is not fitted')
 
-            expected_n_features = self._idf_diag.shape[0]
+            expected_n_features = self._df.shape[0]
             if n_features != expected_n_features:
                 raise ValueError("Input has n_features=%d while the model"
                                  " has been trained with n_features=%d" % (
                                      n_features, expected_n_features))
+
+            idf_diag = sp.spdiags(self.idf_, diags=0, m=n_features,
```

Comment:
Given the power law distribution of vocabularies, this is potentially doing a lot more work at each call to `transform` relative to the vocabulary used in each transformed document. ~~We need a benchmark of repeated calls to `transform`.~~ I now realise that converting the diagonal to CSR in `transform` already requires `O(vocab_size)` time so we're unlikely to see a big change (except that the previous `O(vocab_size)` operation should have been a little faster). :( Perhaps we should be storing `idf_diag` from call to call.

Reply:
So if I understand the issue correctly,
 * `idf_` needs to be exposed and valid after each `partial_fit` since it's a public attribute
 * to calculate `idf_` at each iteration you need the `df` from the previous iteration and the `n_samples`
 * in transform to multiple `X` rows by `idf_` we need to create a diagonal sparse matrix. We can do that either in `partial_fit` (and cache it) or directly in `transform`. The overhead would be the same, and in my opinion, it's largely a matter whether we would a) do numerous `partial_fit` calls b) numerous `transform` calls. I think a) is always true by design (and thus more likely), and I'm not sure if the possible use case of calling `transform` multiple times is worth adding an additional flag for caching `idf_diag`.   

In any case, it might be worth to actually benchmarking the overhead for the creation of `sp.spdiags(self.idf_,..)` to see if it's anything that we need to worry about...

Reply:
@n0mad Aww, so you actually did benchmarks above. Could you also test for e.g. running a transform on a single document?